cherrypick-1.1: improve sentry error reporting

tbg opened the pull request:
Cherry-pick of #19780.

Release note (general change): report the file and line of error causes in anonymized crash reports.

cc @cockroachdb/release

tbg committed:
log: better error reporting

The previous code tended to lose the type when it mattered most, for example in
`errors.Wrap(unknownSentinelErr, "something")`, the output would have only an
empty string for the unknown sentinel error. This will now print its type
instead (which could still be useless, but it's better than nothing).

tbg committed:
log: extract file:line from most opaque errors in reports

This makes previously completely opaque errors such as `errors.Errorf("%s",
"foo")` much more useful by redacting them to their file:line. In particular,
most `errors.Wrap` chains should essentially emit a piecemeal stack trace.

Note that this only works for the `pkg/errors` package, but we use that in most
places (as a drop-in for the `errors` package).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19848)
<!-- Reviewable:end -->

bdarnell commented on pkg/util/log/crash_reporting.go:
```diff
@@ -260,29 +290,22 @@ func reportablesToSafeError(depth int, format string, reportables []interface{})
-		} else {
-			redacted = append(redacted, typ+": "+msg)
-		}
+		s, _ := redact(reportables[i])
```
This is a weird design, with two call sites each of which only uses one of the return values. It's getting really hard to follow this code. Do we really need to keep iterating on this in patch releases?

tschottdorf replied:
If you don't care about getting better sentry error reporting, then no, we don't. But I think we should because a lot of the reports we're getting in sentry are not actionable, and not finding out until next year sets us back.
Re: the design, I'm open for suggestions.

bdarnell replied:
I haven't been paying much attention to sentry so I'm not sure how much we need this. As for design, I can't really suggest a better one because I'm having a hard time telling what we're even trying to do here. Can we reduce some conditional logic by e.g. *always* including the type no matter what?

Basically, my confidence that this code is redacting everything that needs to be redacted is being reduced by every change we add to it. We need to prioritize simplicity here, especially for patch releases.

tbg committed:
log: simplify crash report anonymization

We had an optimization in place that allowed `panic(err)` to report the
type of `err` instead of reporting a wrapper error. This optimization
incurred too much complexity to be worth it and is removed in this commit.

tschottdorf commented:
@bdarnell I cherry-picked #19873 on top of this. See if this is more palatable for 1.1.3.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19848#-:-KyMc-8TB1keebLqhswH:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR!
